core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackStopRestrictionService.java

```diff
@@ -58,19 +95,74 @@ public StopRestrictionReason isInfrastructureStoppable(Stack stack) {
+    }
+
+    private boolean areRequiredServiceRolesPresent(ServiceRoleGroup serviceRoleGroup, Set<ServiceComponent> serviceComponents) {
+        boolean requiredServiceRolesPresent = serviceComponents
+                .containsAll(serviceRolesToServiceComponents(serviceRoleGroup.getRequiredServiceRoles()));
+        boolean requiredRolesPresent = serviceComponents.stream()
+                .map(x -> ServiceComponent.of(null, x.getComponent()))
+                .collect(Collectors.toSet())
+                .containsAll(serviceRolesToServiceComponents(serviceRoleGroup.getRequiredRoles()));
+        return requiredServiceRolesPresent && requiredRolesPresent;
+    }
+
+    private Set<ServiceComponent> serviceRolesToServiceComponents (Set<ServiceRoleGroup.ServiceRole> serviceRoles) {
+        return serviceRoles.stream()
+                .map(x -> ServiceComponent.of(x.getService(), x.getRole()))
+                .collect(Collectors.toSet());
+    }
+
     private boolean hasInstanceGroupAwsEphemeralStorage(InstanceGroup instanceGroup) {
         return instanceGroup.getTemplate().getVolumeTemplates().stream().anyMatch(volume -> AwsDiskType.Ephemeral.value().equals(volume.getVolumeType()));
     }
 
-    private boolean onlyEphemeralStorageInInstanceGroup(InstanceGroup instanceGroup) {
-        return instanceGroup.getTemplate().getVolumeTemplates().stream().allMatch(volume -> AwsDiskType.Ephemeral.value().equals(volume.getVolumeType()));
+    @VisibleForTesting
+    boolean isInstanceGroupEphemeralVolumesOnly (InstanceGroup ig) {
+        long ephemeralVolumeCount = ig.getTemplate().getVolumeTemplates().stream()
+                .filter(volumeTemplate -> AwsDiskType.Ephemeral.value().equalsIgnoreCase(volumeTemplate.getVolumeType())).count();
+        long embeddedDbVolumeCount = ig.getTemplate().getVolumeTemplates().stream()
+                .filter(volumeTemplate -> VolumeUsageType.DATABASE.equals(volumeTemplate.getUsageType())).count();
+        return ephemeralVolumeCount == ig.getTemplate().getVolumeTemplates().size() ||
+                (InstanceGroupType.isGateway(ig.getInstanceGroupType()) &&
+                        ephemeralVolumeCount > 0 &&
+                        ephemeralVolumeCount + embeddedDbVolumeCount == ig.getTemplate().getVolumeTemplates().size());
     }
 
-    private boolean isCbVersionBeforeStopSupport(Stack stack) {
+    private boolean isCbVersionBeforeStopSupport(Stack stack, String minVersion) {
         CloudbreakDetails cloudbreakDetails = componentConfigProviderService.getCloudbreakDetails(stack.getId());
+        return isCbVersionBeforeMinVersion(cloudbreakDetails.getVersion(), minVersion);
+    }
+
+    private boolean isSaltComponentCbVersionBeforeStopSupport(Long clusterId) {
+        String saltCbVersion = clusterComponentProvider.getSaltStateComponentCbVersion(clusterId);
+        return saltCbVersion == null || isCbVersionBeforeMinVersion(saltCbVersion, config.getEphemeralOnlyMinVersion());
+    }
```

Comment:
these seems to be a bit expensive to be called from a `for` loop, as these end in a database select, and these doesn't depend on the `InstanceGroup` at all